cmd/console-proxy/src/posix.rs

```diff
@@ -0,0 +1,401 @@
+        poll_interval: u32,
+    ) -> Result<Self> {
+        let context = HiffyContext::new(hubris, core, hiffy_timeout)?;
+        Ok(Self {
+            hubris,
+            core,
+            context,
+            poll_interval: Duration::from_millis(u64::from(poll_interval)),
+        })
+    }
+
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let op = IdolOperation::new(
+            self.hubris,
+            "ControlPlaneAgent",
+            "uart_read",
+            None,
+        )
+        .context("Could not find `ControlPlaneAgent.uart_read` operation")?;
+
+        let value = humility_cmd_hiffy::hiffy_call(
+            self.hubris,
+            self.core,
+            &mut self.context,
+            &op,
+            &[],
+            Some(HiffyLease::Read(buf)),
+        )?;
+
+        let v = match value {
+            Ok(v) => v,
+            Err(e) => bail!("Got Hiffy error: {e}"),
+        };
+
+        let v =
+            v.as_base()?.as_u32().ok_or_else(|| anyhow!("Couldn't get U32"))?;
+
+        Ok(v as usize)
+    }
+
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        let op = IdolOperation::new(
+            self.hubris,
+            "ControlPlaneAgent",
+            "uart_write",
+            None,
+        )
+        .context("Could not find `ControlPlaneAgent.uart_write` operation")?;
+
+        let buf = &buf[..usize::min(buf.len(), HIFFY_BUF_SIZE)];
+
+        let value = humility_cmd_hiffy::hiffy_call(
+            self.hubris,
+            self.core,
+            &mut self.context,
+            &op,
+            &[],
+            Some(HiffyLease::Write(buf)),
+        )?;
+
+        let v = match value {
+            Ok(v) => v,
+            Err(e) => bail!("Got Hiffy error: {e}"),
+        };
+
+        let v =
+            v.as_base()?.as_u32().ok_or_else(|| anyhow!("Couldn't get U32"))?;
+
+        Ok(v as usize)
+    }
+
+    fn attach(
+        &mut self,
+        raw: bool,
+        imap: RemapRules,
+        omap: RemapRules,
+        log: Option<File>,
+    ) -> Result<()> {
+        // Put terminal in raw mode, if requested, with a guard to restore it.
+        let _guard = if raw {
+            Some(UnrawTermiosGuard::make_stdout_raw()?)
+        } else {
+            None
+        };
+
+        let (stdin_tx, stdin_rx) = crossbeam_channel::unbounded();
+        let (stdout_tx, stdout_rx) = crossbeam_channel::unbounded();
+        thread::spawn(move || stdin_reader(raw, stdin_tx, omap));
+        thread::spawn(move || stdout_writer(stdout_rx, imap, log));
+
+        let mut rx_buf = vec![0; HIFFY_BUF_SIZE];
+        let mut tx_buf = Vec::new();
+        loop {
+            if !tx_buf.is_empty() {
+                let nwritten = self.write(&tx_buf)?;
+                tx_buf.drain(0..nwritten);
+            }
+
+            let nread = self.read(&mut rx_buf)?;
+            if nread > 0 {
+                stdout_tx.send(rx_buf[..nread].to_owned()).unwrap();
```

Comment:
Why does this need to be in a separate thread, versus writing `rx_buf[..nread]` to `stdout` right here?

Reply:
I don't think it does! When I originally wrote this I was thinking that we'd offload I/O to other threads so we could quickly service repeated reads/writes, but, uh, yeah, not so much. I'll get rid of the stdout thread.